wrappers/services.go

```diff
@@ -375,120 +379,270 @@ func userDaemonReload() error {
 	return cli.ServicesDaemonReload(ctx)
 }
 
-// AddSnapServicesOptions is a struct for controlling the generated service
-// definition for a snap service.
-type AddSnapServicesOptions struct {
-	Preseeding              bool
-	VitalityRank            int
+type SnapServiceOptions struct {
+	// VitalityRank is the rank of all services in the specified snap used by
+	// the OOM killer when OOM conditions are reached.
+	VitalityRank int
+}
+
+// EnsureSnapServicesOptions is the set of options applying to the
+// EnsureSnapServices operation. It does not include per-snap specific options
+// such as VitalityRank or RequireMountedSnapdSnap from AddSnapServiceOptions,
+// since those are expected to be provided in the snaps argument.
+type EnsureSnapServicesOptions struct {
+	// Preseeding is whether the system is currently being preseeded, in which
+	// case there is not a running systemd for EnsureSnapServicesOptions to
+	// issue commands like systemctl daemon-reload to.
+	Preseeding bool
+
+	// RequireMountedSnapdSnap is whether the generated units should depend on
+	// the snapd snap being mounted, this is specific to systems like UC18 and
+	// UC20 which have the snapd snap and need to have units generated
 	RequireMountedSnapdSnap bool
 }
 
-// AddSnapServices adds service units for the applications from the snap which
-// are services. The services do not get enabled or started.
-func AddSnapServices(s *snap.Info, opts *AddSnapServicesOptions, inter interacter) (err error) {
-	if s.Type() == snap.TypeSnapd {
-		return fmt.Errorf("internal error: adding explicit services for snapd snap is unexpected")
-	}
+// EnsureSnapServices will ensure that the specified snap services' file states
+// are up to date with the specified options and infos. It will add new services
+// if those units don't already exist, but it does not delete existing service
+// units that are not present in the snap's Info structures.
+// There are two sets of options; there are global options which apply to the
+// entire transaction and to every snap service that is ensured, and options
+// which are per-snap service and specified in the map argument.
+// The return value is a map of the snap's Info the a list of App's that were
+// updated or modified. Apps in a snap that did not have their service
+// definitions change are not included in the map's list value.
+// If any errors are encountered trying to update systemd units, then all
+// changes performed up to that point are rolled back, meaning newly written
+// units are deleted and modified units are attempted to be restored to their
+// previous state. In this case the modified return value is nil.
+func EnsureSnapServices(snaps map[*snap.Info]*SnapServiceOptions, opts *EnsureSnapServicesOptions, inter interacter) (modified map[*snap.Info][]*snap.AppInfo, err error) {
```

Comment:
hmm this still has the modified return value, I would have expected this PR to also have included this PR from @pedronis: https://github.com/snapcore/snapd/pull/10176

Reply:
Indeed, I updated the PR to include the other two cherry picks here for easier review

Reply:
this PR is still missing https://github.com/snapcore/snapd/commit/e7f02ea82b60149edaa08378956e8be06d9bac5f